ledger/smt.go

```diff
@@ -0,0 +1,402 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ledger
+
+// The SMT is derived from https://github.com/aergoio/SMT with modifications
+// to remove unneeded features, and to support retention of old nodes for a fixed time.
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"time"
+
+	"istio.io/pkg/cache"
+)
+
+// TODO when using the SMT, make sure keys and values are same length as Hash
+
+// SMT is a sparse Merkle tree.
+type SMT struct {
+	// Root is the current root of the smt.
+	Root []byte
```

Comment:
If this implementation is hidden and only used by the ledger abstraction, would it be simpler to change the []byte stuff here and elsewhere into strings, thus avoiding all the casting in ledger.go?

Reply:
The SMT operates on keys one bit at a time, and it is not clear when an array is initialized whether a given element will contain a key or a value.  Storing strings would mean casting them to bytes at read/write time in order to properly store and retrieve them, including intermediate nodes, which are always bytes.  I think casting at the edge is likely the right thing to do here...